File: python/020.py
#   020
#   Valid Parentheses
#   2015-04-28
#####################################################
class Solution:
    # @param {string} s
    # @return {boolean}
    def isValid(self, s):
        stack = []
        for c in s:
            if not stack:
                stack.append(c)
            elif (c == ')' and stack[-1] == '(') or (c == ']' and stack[-1] == '[') or (c == '}' and stack[-1] == '{'):
                stack.pop()
            else:
                stack.append(c)
        return not stack

#####################################################

s = "]"
print(s)
sol = Solution()
z = sol.isValid(s)
print("True" if z else "False")

File: ruby/020.rb
# 020
# Valid Parentheses
# 2015-04-28
#####################################################
# @param {String} s
# @return {Boolean}
def is_valid(s)
  stack = []
  s.each_char do |c|
    if (c == ')' and stack.last == '(') or (c == ']' and stack.last == '[') or (c == '}' and stack.last == '{')
      stack.pop
    else
      stack.push(c)
    end
  end
  stack.empty?
end

#####################################################
s = "]"
puts s
puts is_valid(s) ? "True" : "False"

